# Análise Imobiliária em Portugal - Semana 5

## 📊 Objetivo
Explorar e visualizar padrões de preços de imóveis em Portugal usando dataset com 135k+ listagens, 
identificando oportunidades de investimento por região, tipo de imóvel e certificado energético.

## ❓ Perguntas a Responder
- Qual é a distribuição de preços por distrito português?
- Como o certificado energético influencia o valor do imóvel?
- Qual é o preço/m² em Lisboa vs outras regiões?
- Que tipo de imóvel (apartamento vs casa) oferece melhor custo-benefício?
- Existe relação entre ano de construção e preço?

## 📈 Dataset
- **Fonte:** Kaggle - Real Estate Listings in Portugal 2024
- **Linhas:** 135.236 imóveis
- **Colunas:** 25 atributos (preço, localização, tipo, certificado energético, características físicas)
- **Período:** Listagens atualizadas semanalmente

## 🛠️ Ferramentas
- Python (Pandas, NumPy)
- Visualizações (Plotly, Matplotlib)
- Análise exploratória (EDA)

## 🎯 Resultados Esperados
1. Dashboard interativo em Streamlit
2. Insights de mercado por região
3. Recomendações de investimento baseadas em dados
4. Código reproduzível no GitHub

## 📝 Etapas
1. **Carregamento & Limpeza:** Remover outliers, tratar nulos, validar tipos
2. **Exploração:** Distribuições, correlações, estatísticas
3. **Visualização:** Gráficos interativos por district, energy cert, type
4. **Dashboard:** App Streamlit com filtros
5. **Deploy:** GitHub + LinkedIn post

---


# Bibliotecas

In [1]:
import pandas as pd
import numpy as np
import plotly.express as px
import plotly.graph_objects as go
from plotly.subplots import make_subplots
import warnings
warnings.filterwarnings('ignore')

## Carregamento dos dados

In [3]:
df_raw = pd.read_csv("../data/raw/portugal_listinigs.csv")

In [4]:
#Formato do dataframe (shape)
print(f"O dataframe possui {df_raw.shape[0]} linhas e {df_raw.shape[1]} colunas. ")

#Colunas do dataframe
print(f"\nAs colunas do dataframe são: {df_raw.columns.tolist()}.")

#Colunas e sua posição
print("\nColunas e suas posições:")
for i, col in enumerate(df_raw.columns):
    print(f"{i}: {col}")    


O dataframe possui 135536 linhas e 25 colunas. 

As colunas do dataframe são: ['Price', 'District', 'City', 'Town', 'Type', 'EnergyCertificate', 'GrossArea', 'TotalArea', 'Parking', 'HasParking', 'Floor', 'ConstructionYear', 'EnergyEfficiencyLevel', 'PublishDate', 'Garage', 'Elevator', 'ElectricCarsCharging', 'TotalRooms', 'NumberOfBedrooms', 'NumberOfWC', 'ConservationStatus', 'LivingArea', 'LotSize', 'BuiltArea', 'NumberOfBathrooms'].

Colunas e suas posições:
0: Price
1: District
2: City
3: Town
4: Type
5: EnergyCertificate
6: GrossArea
7: TotalArea
8: Parking
9: HasParking
10: Floor
11: ConstructionYear
12: EnergyEfficiencyLevel
13: PublishDate
14: Garage
15: Elevator
16: ElectricCarsCharging
17: TotalRooms
18: NumberOfBedrooms
19: NumberOfWC
20: ConservationStatus
21: LivingArea
22: LotSize
23: BuiltArea
24: NumberOfBathrooms


In [5]:
#Pevisualização dos dados
df_raw.head()

,Price,District,City,Town,Type,EnergyCertificate,GrossArea,TotalArea,Parking,HasParking,...,Elevator,ElectricCarsCharging,TotalRooms,NumberOfBedrooms,NumberOfWC,ConservationStatus,LivingArea,LotSize,BuiltArea,NumberOfBathrooms
0,780000.0,Vila Real,Valpaços,Carrazedo de Montenegro e Curros,Farm,NC,200.0,552450.0,0.0,False,...,False,NaN,NaN,NaN,NaN,NaN,120.0,NaN,NaN,0.0
1,223000.0,Faro,São Brás de Alportel,São Brás de Alportel,Apartment,A+,NaN,81.0,1.0,True,...,True,NaN,2.0,NaN,NaN,NaN,81.0,NaN,NaN,2.0
2,228000.0,Faro,São Brás de Alportel,São Brás de Alportel,Apartment,A+,NaN,108.0,1.0,True,...,True,NaN,2.0,NaN,NaN,NaN,108.0,NaN,NaN,2.0
3,250000.0,Faro,São Brás de Alportel,São Brás de Alportel,Apartment,A+,NaN,114.0,1.0,True,...,True,NaN,2.0,NaN,NaN,NaN,114.0,NaN,NaN,0.0
4,250000.0,Faro,São Brás de Alportel,São Brás de Alportel,Apartment,A+,NaN,114.0,1.0,True,...,True,NaN,2.0,NaN,NaN,NaN,114.0,NaN,NaN,2.0


In [6]:
# Information do dataframe
print("\nInformações do dataframe:")
df_raw.info()


Informações do dataframe:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 135536 entries, 0 to 135535
Data columns (total 25 columns):
 #   Column                 Non-Null Count   Dtype  
---  ------                 --------------   -----  
 0   Price                  135236 non-null  float64
 1   District               135536 non-null  object 
 2   City                   135536 non-null  object 
 3   Town                   135534 non-null  object 
 4   Type                   135520 non-null  object 
 5   EnergyCertificate      135522 non-null  object 
 6   GrossArea              27638 non-null   float64
 7   TotalArea              127153 non-null  float64
 8   Parking                135342 non-null  float64
 9   HasParking             68215 non-null   object 
 10  Floor                  27929 non-null   object 
 11  ConstructionYear       88021 non-null   float64
 12  EnergyEfficiencyLevel  67289 non-null   object 
 13  PublishDate            29239 non-null   object 
 14  Garage   

In [7]:
# estatísticas descritivas
print("\nEstatísticas descritivas:")
df_raw.describe()


Estatísticas descritivas:


,Price,GrossArea,TotalArea,Parking,ConstructionYear,TotalRooms,NumberOfBedrooms,NumberOfWC,LivingArea,LotSize,BuiltArea,NumberOfBathrooms
count,1.352360e+05,2.763800e+04,1.271530e+05,135342.000000,88021.000000,73244.000000,47041.000000,57256.000000,1.049520e+05,3.958300e+04,2.661700e+04,128700.000000
mean,3.681374e+05,2.789733e+03,5.140780e+05,0.549667,1989.303825,3.229357,2.665823,0.430785,1.403037e+03,5.636939e+04,3.086621e+03,1.544025
std,3.804044e+06,1.153670e+05,1.722937e+08,0.839257,26.684309,10.516977,1.790474,1.016360,3.474597e+04,5.208474e+06,1.185137e+05,1.764637
min,1.000000e+00,-7.000000e+00,-7.196067e+06,0.000000,1900.000000,0.000000,0.000000,-15.000000,0.000000e+00,0.000000e+00,-1.000000e+00,-13.000000
25%,8.400000e+04,1.000000e+02,9.100000e+01,0.000000,1973.000000,2.000000,2.000000,0.000000,8.000000e+01,2.580000e+02,1.045700e+02,0.000000
50%,2.100000e+05,1.640000e+02,1.590000e+02,0.000000,1994.000000,3.000000,3.000000,0.000000,1.180000e+02,6.790000e+02,1.680000e+02,1.000000
75%,3.950000e+05,2.940000e+02,5.000000e+02,1.000000,2009.000000,4.000000,3.000000,1.000000,2.040000e+02,2.890000e+03,3.000000e+02,2.000000
max,1.380000e+09,1.275000e+07,6.142007e+10,3.000000,2025.000000,2751.000000,21.000000,59.000000,5.429000e+06,9.923010e+08,1.275000e+07,131.000000


## Análise de Dados Faltantes (Missing Data)

In [9]:
# Percentual de nulos por coluna
missing = (df_raw.isnull().sum() / len(df_raw) * 100).sort_values(ascending=False)
print("Dados Faltantes (%):")
print(missing[missing > 0])

# Visualiza
fig = px.bar(
    x=missing[missing > 0].values,
    y=missing[missing > 0].index,
    orientation='h',
    title='Percentual de Dados Faltantes',
    labels={'x': '% Nulo', 'y': 'Coluna'}
)
fig.show()

Dados Faltantes (%):
ConservationStatus       85.766143
BuiltArea                80.361675
GrossArea                79.608370
Floor                    79.393667
PublishDate              78.427134
LotSize                  70.795213
NumberOfBedrooms         65.292616
NumberOfWC               57.755873
EnergyEfficiencyLevel    50.353412
ElectricCarsCharging     50.353412
Garage                   50.353412
HasParking               49.670198
TotalRooms               45.959745
ConstructionYear         35.057107
LivingArea               22.565223
TotalArea                 6.185073
NumberOfBathrooms         5.043678
Price                     0.221343
Parking                   0.143135
Elevator                  0.023610
Type                      0.011805
EnergyCertificate         0.010329
Town                      0.001476
dtype: float64


## Limpeza de Dados

### Problemas Identificados:
- Price: min=1€, max=1.38B€ (outliers extremos)
- TotalArea: valores negativos e absurdos (992M m²)
- NumberOfBedrooms: máximo 21 (possivéis erros de entrada)
- Mixed types em 8 colunas (conversão necessária)

### Estratégia:
1. Remove preços < 10k€ e > 5M€
2. Remove áreas negativas ou > 100k m²
3. Remove dormitórios > 10
4. Cria coluna Price/m²

In [8]:
# Cria cópia independente para limpeza
df_clean = df_raw.copy()

print(f"Linhas antes: {len(df_clean)}")

# 1. Preços: 10k a 5M
df_clean = df_clean[(df_clean['Price'] >= 10000) & (df_clean['Price'] <= 5000000)]

# 2. Área total: > 0 e < 100k m²
df_clean = df_clean[(df_clean['TotalArea'] > 0) & (df_clean['TotalArea'] <= 100000)]

# 3. Dormitórios: 0-10
df_clean = df_clean[(df_clean['NumberOfBedrooms'] >= 0) & (df_clean['NumberOfBedrooms'] <= 10)]

# 4. Remove linhas com Price ou TotalArea nulos (essenciais)
df_clean = df_clean.dropna(subset=['Price', 'TotalArea'])

print(f"Linhas após limpeza: {len(df_clean)}")
print(f"Linhas removidas: {len(df_raw) - len(df_clean)} ({(len(df_raw) - len(df_clean))/len(df_raw)*100:.1f}%)")

Linhas antes: 135536
Linhas após limpeza: 45713
Linhas removidas: 89823 (66.3%)


In [10]:
# Cria coluna Price/m²
df_clean['PricePerM2'] = df_clean['Price'] / df_clean['TotalArea']

print(df_clean[['Price', 'TotalArea', 'PricePerM2']].describe())

              Price     TotalArea     PricePerM2
count  4.571300e+04  45713.000000   45713.000000
mean   3.920106e+05    519.822392    2831.613224
std    4.417480e+05   3409.764856    2551.344233
min    1.000000e+04      1.000000       0.252620
25%    1.650000e+05     84.000000    1363.636364
50%    2.799000e+05    119.000000    2403.669725
75%    4.480000e+05    189.000000    3774.157303
max    5.000000e+06  99000.000000  260000.000000


## 📊 Análise Exploratória (EDA)

### Distribuição de Preços

In [11]:
fig = px.histogram(
    df_clean,
    x='Price',
    nbins=50,
    title='Distribuição de Preços de Imóveis',
    labels={'Price': 'Preço (€)'},
    color_discrete_sequence=['#1f77b4']
)
fig.update_layout(showlegend=False)
fig.show()

print(f"Preço Mediano: €{df_clean['Price'].median():,.0f}")
print(f"Preço Médio: €{df_clean['Price'].mean():,.0f}")

Preço Mediano: €279,900
Preço Médio: €392,011


### Preços por Tipo de Imóvel

In [22]:
fig = px.box(
    df_clean,
    x='Type',
    y='Price',
    title='Distribuição de Preços por Tipo de Imóvel',
    labels={'Price': 'Preço (€)', 'Type': 'Tipo'},
    color='Type'
)
fig.show()

print("\nPreço Médio por Tipo:")
print(df_clean.groupby('Type')['Price'].agg(['mean', 'median', 'count']).sort_values('median', ascending=False))

# Calcula premium de Casa vs Apartamento
if 'House' in df_clean['Type'].values and 'Apartment' in df_clean['Type'].values:
    house_price = df_clean[df_clean['Type'] == 'House']['Price'].median()
    apartment_price = df_clean[df_clean['Type'] == 'Apartment']['Price'].median()
    premium = ((house_price - apartment_price) / apartment_price) * 100
    
    print(f"\nCasa vs Apartamento:")
    print(f"Casa mediana: €{house_price:,.0f}")
    print(f"Apartamento mediano: €{apartment_price:,.0f}")
    print(f"Premium Casa: {premium:.1f}% {'mais caro' if premium > 0 else 'mais barato'}")


Preço Médio por Tipo:
                             mean     median  count
Type                                               
Manor                2.440000e+06  2440000.0      2
Mansion              1.893571e+06  1500000.0      7
Estate               1.725825e+06  1300000.0     12
Hotel                1.434477e+06  1200000.0     13
Industrial           1.169561e+06   750000.0     27
Duplex               7.336551e+05   500000.0    185
Building             5.825851e+05   400000.0    447
Farm                 5.913858e+05   349950.0   1002
Apartment            3.856594e+05   298500.0  23319
House                4.075862e+05   265000.0  17758
Warehouse            4.497765e+05   212000.0     85
Other - Commercial   4.841677e+05   210000.0     51
Office               4.637019e+05   192500.0     52
Investment           4.022627e+05   185000.0     51
Studio               2.079734e+05   180000.0     47
Store                2.328624e+05   125000.0    419
Transfer of lease    1.670959e+05    8000

### Preços por Distrito (Top 10)

In [13]:
district_stats = df_clean.groupby('District').agg({
    'Price': ['median', 'mean', 'count']
}).round(0)
district_stats.columns = ['Mediana', 'Média', 'Qtd_Imóveis']
district_stats = district_stats.sort_values('Mediana', ascending=False).head(10)

fig = px.bar(
    district_stats.reset_index(),
    x='District',
    y='Mediana',
    title='Top 10 Distritos - Preço Mediano',
    labels={'Mediana': 'Preço Mediano (€)', 'District': 'Distrito'},
    color='Mediana',
    color_continuous_scale='Viridis'
)
fig.update_xaxes(tickangle=-45)
fig.show()

print(district_stats)

                       Mediana     Média  Qtd_Imóveis
District                                             
Z - Fora de Portugal  535000.0  531667.0           21
Lisboa                399900.0  598654.0        11475
Faro                  370000.0  541600.0         3100
Setúbal               335000.0  440181.0         4133
Porto                 325000.0  370500.0         8765
Ilha Terceira         320000.0  361846.0           13
Ilha de São Miguel    249900.0  350815.0           17
Braga                 240000.0  292714.0         3633
Ilha de Porto Santo   235000.0  199700.0            6
Aveiro                235000.0  285024.0         2153


### Influência do Certificado Energético

In [20]:
energy_stats = df_clean[df_clean['EnergyCertificate'].notna()].groupby('EnergyCertificate').agg({
    'Price': ['median', 'mean', 'count'],
    'PricePerM2': ['median', 'mean']
}).round(0)

fig = px.box(
    df_clean[df_clean['EnergyCertificate'].notna()],
    x='EnergyCertificate',
    y='Price',
    title='Preço por Certificado Energético',
    labels={'Price': 'Preço (€)', 'EnergyCertificate': 'Certificado'},
    color='EnergyCertificate'
)
fig.show()

print("Preço Mediano por Certificado Energético:")
print(energy_stats)

# Calcula premium A+
if ('A+',) in energy_stats.index.get_level_values(0) or 'A+' in energy_stats.index:
    a_plus_price = df_clean[df_clean['EnergyCertificate'] == 'A+']['Price'].median()
    general_price = df_clean['Price'].median()
    premium = ((a_plus_price - general_price) / general_price) * 100
    print(f"\n✨ Premium A+: {premium:.1f}% (€{a_plus_price:,.0f} vs €{general_price:,.0f})")

Preço Mediano por Certificado Energético:
                      Price                  PricePerM2        
                     median      mean  count     median    mean
EnergyCertificate                                              
A                  430000.0  607259.0   4833     3415.0  4002.0
A+                 422500.0  560776.0   2330     2907.0  3312.0
B                  437500.0  634341.0   1806     3254.0  3952.0
B-                 399000.0  583160.0   1603     3028.0  3799.0
C                  290000.0  418827.0   6249     2640.0  3115.0
D                  249500.0  358448.0   6663     2397.0  2827.0
E                  215000.0  306966.0   4842     2000.0  2364.0
F                  155000.0  238164.0   3336     1283.0  1766.0
G                  130000.0  214777.0    112     1000.0  1547.0
NC                 225000.0  307900.0  13901     1847.0  2393.0
No Certificate     171000.0  196166.0     38     1842.0  2012.0

✨ Premium A+: 50.9% (€422,500 vs €279,900)


### Preço/m² por Distrito (Braga destacada)

In [15]:
priceperm2_stats = df_clean.groupby('District')['PricePerM2'].agg(['median', 'mean', 'count']).sort_values('median', ascending=False).head(10)

fig = px.bar(
    priceperm2_stats.reset_index(),
    x='District',
    y='median',
    title='Top 10 - Preço/m² por Distrito',
    labels={'median': 'Preço/m² (€)', 'District': 'Distrito'},
    color='median',
    color_continuous_scale='RdYlGn_r'
)
fig.update_xaxes(tickangle=-45)
fig.show()

# Foco Braga
if 'Braga' in df_clean['District'].values:
    braga_price = df_clean[df_clean['District'] == 'Braga']['PricePerM2'].median()
    portugal_price = df_clean['PricePerM2'].median()
    print(f"\nBraga: €{braga_price:.0f}/m²")
    print(f"Portugal (média): €{portugal_price:.0f}/m²")
    print(f"Diferença: {((braga_price / portugal_price - 1) * 100):.1f}% mais caro")


Braga: €1667/m²
Portugal (média): €2404/m²
Diferença: -30.7% mais caro


### Correlações Entre Variáveis

In [16]:
# Seleciona colunas numéricas para correlação
numeric_cols = df_clean[['Price', 'TotalArea', 'NumberOfBedrooms', 'NumberOfWC', 'PricePerM2', 'ConstructionYear']].dropna()

corr_matrix = numeric_cols.corr()

fig = px.imshow(
    corr_matrix,
    title='Matriz de Correlação',
    labels=dict(color='Correlação'),
    color_continuous_scale='RdBu_r',
    zmin=-1,
    zmax=1
)
fig.show()

print("\nCorrelações com Preço:")
print(corr_matrix['Price'].sort_values(ascending=False))


Correlações com Preço:
Price               1.000000
PricePerM2          0.526691
NumberOfBedrooms    0.365261
ConstructionYear    0.258841
NumberOfWC          0.170231
TotalArea          -0.008242
Name: Price, dtype: float64


## 💡 Insights Principais

1. **Distribuição de Preços**: Maioria concentrada em €84k-€395k (Q1-Q3)
2. **Tipo de Imóvel**: Casas custam ~[valor] mais que apartamentos
3. **Localização**: Braga é **30,7% mais barato** que a média nacional
4. **Certificado Energético**: Imóveis A+ têm 50.9%% de premium
5. **Espaço/Preço**: Forte correlação entre TotalArea e Price

In [18]:
# Resumo final
print("=" * 60)
print("RESUMO DA ANÁLISE")
print("=" * 60)
print(f"Total de imóveis analisados: {len(df_clean):,}")
print(f"Faixa de preço: €{df_clean['Price'].min():,.0f} - €{df_clean['Price'].max():,.0f}")
print(f"Preço mediano: €{df_clean['Price'].median():,.0f}")
print(f"Preço/m² mediano: €{df_clean['PricePerM2'].median():,.0f}")
print(f"\nTipos de imóvel: {df_clean['Type'].nunique()}")
print(f"Distritos: {df_clean['District'].nunique()}")
print("=" * 60)


RESUMO DA ANÁLISE
Total de imóveis analisados: 45,713
Faixa de preço: €10,000 - €5,000,000
Preço mediano: €279,900
Preço/m² mediano: €2,404

Tipos de imóvel: 21
Distritos: 26
